[IDE][Index] Property wrapper rename support

nathawes opened the pull request:
This PR fixes up rename support for property wrappers. There were three main issues this addresses:

1. Renaming an `@propertyWrapper` struct was failing due to rename's `NameMatcher` using the range of a `Decl` itself, not including its attributes, to decide if it contained a location that needed to be resolved.
2. Renaming a constructor of a `@propertyWrapper` struct would not pick up occurrences of that constructor's argument labels used in custom attributes, e.g. renaming `MyWrapper`'s `init(initialValue:minValue:maxValue)` constructor would miss the argument labels in custom attributes like `@MyWrapper(initialValue: 50, minValue: 0, maxValue: 100)`. This is because the indexer didn't report an occurrence of the constructor call in custom attributes. Rename's `NameMatcher` also needed to be updated to handle resolving these newly-added occurrences, since the call expression in the AST doesn't have an explicit base (`MyWrapper` is not part of the call expression, but is where the index reports the call taking place).
3. Renaming a wrapped property (e.g. `foo`) wasn't updating the corresponding synthesized `$`-prefixed backing property (`$foo`). The synthesized backing property declarations were implicit, so the index wasn't recording them, and even if it did, it's a different property so wouldn't be picked up automatically.
\
To resolve this, the index now records an occurrence of the wrapped property within the synthesized backing property occurrences (i.e. on the `foo` within each `$foo`) so that rename on the wrapped property updates the synthesized backing property occurrences too.  To support rename being invoked via an occurrence of a synthesized property, the `CusorInfo` request that determines which symbol is renamed for a given location now treats occurrences of synthesized properties as if they were the corresponding wrapped properties. The `FindRelatedIdents` request was also similarly updated so that rename-like features that are based on it (like Xcode's edit-all-in-scope) can get both the underlying and synthesized property ranges, regardless of which is was invoked on.

This PR also includes some small and/or incidental IDE issue fixes around custom attributes:
- syntax coloring wasn't working for the custom attribute type occurrences (only semantic highlighting)
- interleaving of availability attributes, custom attributes and built-in attributes would sometimes result in incorrect syntax highlighting.

Resolves rdar://problem/49036613
Resolves rdar://problem/50073641
Resolves rdar://problem/51695783
Resolves rdar://problem/52053678

Commit:
[IDE][Index] Fix syntax coloring, index, and rename support for custom attributes

This fixes custom attribute syntax highlighting on parameters and functions
(where function builders can be applied). They weren't being walked in
the function position previously and were walked out of source order in the
parameter position.

It also fixes rename of the property wrapper and function builder type
names that can appear in custom attributes, as well as rename of property
wrapper constructors, that can appear after the type names, e.g.
`@Wrapper(initialValue: 10)`. The index now also records these constructor
occurrences, along with implicit occurrences whenever a constructor is
called via default value assignment, e.g. `@Wrapper var foo = 10`, so that
finding calls/references to the constructor includes these locations.

Resolves rdar://problem/49036613
Resolves rdar://problem/50073641

Commit:
[IDE][Index] Renaming a wrapped property should also rename the synthesized $-prefixed backing property

This patch achieves this by updating indexing to reporting the position of
`foo` in occurrences of `$foo` as an occurrence of the `foo` symbol, so
that renames initiated on occurrences of the `foo` symbol will also result
in occurrences of the `$foo` symbol being updated correctly. This also means
find-references on foo will show places where $foo is used.

Making rename work in the other direction (invoking rename on $foo upating foo
occurrences too) is still todo.

Commit:
[IDE] Update CursorInfo and FindRelatedIdents to treat property wrapper backing properties as the underlying property

This is the final piece to get rename and findRelatedIdents to include all foo
and $foo occurrences, for example, in the set of occurrences to be renamed or
shown, regardless of whether they were initiated on a foo or $foo.

Resolves rdar://problem/51695783.

Commit:
[test] Update Index/refactoring property wrapper tests to use wrappedValue rather than value

Plus other small cleanups to comments and variable names.

nathawes commented:
@swift-ci please test

nathawes commented:
@swift-ci please test

benlangmuir commented on lib/IDE/SwiftSourceDocInfo.cpp:
```diff
@@ -354,6 +354,39 @@ static std::vector<CharSourceRange> getEnumParamListInfo(SourceManager &SM,
 
+bool NameMatcher::handleCustomAttrs(Decl *D) {
+  // CustomAttrs of non-param VarDecls are handled by their containing
+  // PatternBindingDecl
```
Should add this comment to `SemaAnnotator::handleCustomAttributes` as well.

benlangmuir commented on lib/Index/Index.cpp:
```diff
@@ -450,6 +477,16 @@ class IndexSwiftASTWalker : public SourceEntityWalker {
 
+  Decl *getContainingDecl() const {
+    auto Containers = (SymbolRoleSet)SymbolRole::Definition | (SymbolRoleSet)SymbolRole::Declaration;
+    for (const auto &Entity: EntitiesStack) {
```
When does a declaration (not a definition) contain something?  Is this for default argument expressions or something?

nathawes replied:
Nope, this just doesn't make sense 🙂
I don't even think we use the Declaration role in swift...

benlangmuir commented on test/Index/property_wrappers.swift:
```diff
@@ -18,25 +18,38 @@ var globalInt: Int { return 17 }
   @Wrapper
   public var x: Int = globalInt
+  // CHECK: [[@LINE-2]]:4 | struct/Swift | Wrapper | s:14swift_ide_test7WrapperV | Ref | rel: 0
+  // CHECK-NOT: [[@LINE-2]]:23 | variable/Swift | globalInt | s:14swift_ide_test9globalIntSivp | Ref,Read | rel: 0
```
I suggest dropping everything after the name `globalInt` to make this CHECK-NOT more robust against changes.

benlangmuir commented on test/refactoring/SyntacticRename/FindRangeOutputs/property-wrapper-init/body.swift.expected:
```diff
@@ -0,0 +1,37 @@
+    @propertyWrapper
+    public struct InnerWrapper<T> {
+        public var value: T
+        public /*init:def*/init(initialValue: T) {
```
Do you have a test somewhere for `init(initialValue: T, other: T)` with `@Wrapper(other: 1) var foo: Int = 2`?

nathawes replied:
Oh thanks – I didn't realize that was in yet. I think the initializer rename should still be fine, but given the passed arguments could be closure bodies and can now be split between the custom attribute and the var initializer, NameMatcher and SyntaxModel may have issues with things being out of source order...

nathawes replied:
Added tests and it looks like they're actually fine. They're running over the non-typechecked AST, so they're still two distinct pieces.

Commit:
[test] Add more index/refactoring/cursor-info tests for property wrappers

Make sure they handle the case when a property wrapper type's constructor is
called with the first argument coming from the var initializer, and the rest
from the custom attribute's argument.

nathawes commented:
@swift-ci please test

Commit:
[IDE][Index][test] Update sourcekit/indexing support for latest property wrapper changes

The backing property for 'foo' is now '_foo', and the projected value '$foo'.
This updates Indexing to report occurrences of foo within both $foo and
_foo occurrences (rather than just $foo - the old _foo).

FindRelatedIdents was similarlar updated, so it reports 'foo' ranges in both
_foo and $foo.

CursorInfo now reports the USR, documentation, and location of foo when invoked
occurrences of $foo or _foo, but now leaves the name, type, and annotated
declaration of _foo/$foo as is. Having the same USR ensures rename invoked on
any of them will still rename via foo. Reporting foo's documentation comment
instead is just to present something more useful to the user.

nathawes commented:
@swift-ci please test

nathawes commented:
@swift-ci please test

nathawes commented:
@swift-ci please test

nathawes commented:
@swift-ci please test OS X Platform
